File: src/data/screener.ts
import YahooFinance from "yahoo-finance2"

export type ScreenId =
  | "aggressive_small_caps"
  | "conservative_foreign_funds"
  | "day_gainers"
  | "day_losers"
  | "growth_technology_stocks"
  | "high_yield_bond"
  | "most_actives"
  | "most_shorted_stocks"
  | "portfolio_anchors"
  | "small_cap_gainers"
  | "solid_large_growth_funds"
  | "solid_midcap_growth_funds"
  | "top_mutual_funds"
  | "undervalued_growth_stocks"
  | "undervalued_large_caps"

export const SCREEN_IDS: ScreenId[] = [
  "aggressive_small_caps",
  "conservative_foreign_funds",
  "day_gainers",
  "day_losers",
  "growth_technology_stocks",
  "high_yield_bond",
  "most_actives",
  "most_shorted_stocks",
  "portfolio_anchors",
  "small_cap_gainers",
  "solid_large_growth_funds",
  "solid_midcap_growth_funds",
  "top_mutual_funds",
  "undervalued_growth_stocks",
  "undervalued_large_caps",
]

export interface ScreenedQuote {
  symbol: string
  name: string
  marketCap?: number
  peRatio?: number
  forwardPe?: number
  price?: number
  changePercent?: number
}

const yf = new YahooFinance({ suppressNotices: ["yahooSurvey"] })

export async function runScreen(scrId: ScreenId, count = 25): Promise<ScreenedQuote[]> {
  try {
    const res = await yf.screener({ scrIds: scrId, count })
    return res.quotes.map((q) => {
      const out: ScreenedQuote = {
        symbol: q.symbol,
        name: q.shortName ?? q.longName ?? q.symbol,
      }
      if (q.marketCap != null) out.marketCap = q.marketCap
      if (q.trailingPE != null) out.peRatio = q.trailingPE
      if (q.forwardPE != null) out.forwardPe = q.forwardPE
      if (q.regularMarketPrice != null) out.price = q.regularMarketPrice
      if (q.regularMarketChangePercent != null) out.changePercent = q.regularMarketChangePercent
      return out
    })
  } catch {
    return []
  }
}

File: app/data/screener.py
"""Yahoo predefined equity screeners via yfinance (replaces yahoo-finance2 screener)."""

from __future__ import annotations

import yfinance as yf
from pydantic import BaseModel

# The predefined Yahoo screen keys, mirrored from the original ScreenId union.
SCREEN_IDS: list[str] = [
    "aggressive_small_caps",
    "conservative_foreign_funds",
    "day_gainers",
    "day_losers",
    "growth_technology_stocks",
    "high_yield_bond",
    "most_actives",
    "most_shorted_stocks",
    "portfolio_anchors",
    "small_cap_gainers",
    "solid_large_growth_funds",
    "solid_midcap_growth_funds",
    "top_mutual_funds",
    "undervalued_growth_stocks",
    "undervalued_large_caps",
]


class ScreenedQuote(BaseModel):
    symbol: str
    name: str
    market_cap: float | None = None
    pe_ratio: float | None = None
    forward_pe: float | None = None
    price: float | None = None
    change_percent: float | None = None


def run_screen(scr_id: str, count: int = 25) -> list[ScreenedQuote]:
    try:
        res = yf.screen(scr_id, count=count)
    except Exception:
        return []
    quotes = (res or {}).get("quotes", []) if isinstance(res, dict) else []
    out: list[ScreenedQuote] = []
    for q in quotes:
        sym = q.get("symbol")
        if not sym:
            continue
        out.append(
            ScreenedQuote(
                symbol=sym,
                name=q.get("shortName") or q.get("longName") or sym,
                market_cap=q.get("marketCap"),
                pe_ratio=q.get("trailingPE"),
                forward_pe=q.get("forwardPE"),
                price=q.get("regularMarketPrice"),
                change_percent=q.get("regularMarketChangePercent"),
            )
        )
    return out
